performance: Speed up Method Completions By Taking Advantage of Orphan Rules

davidbarsky opened the pull request:
(Continues https://github.com/rust-lang/rust-analyzer/pull/16498)

This PR speeds up method completions by doing two things without regressing `analysis-stats`[^1]:
- Filter candidate traits prior to calling `iterate_path_candidates` by relying on orphan rules (see below for a slightly more in-depth explanation). When generating completions [on `slog::Logger`](https://github.com/oxidecomputer/omicron/blob/5e9e59c312cd787ba50cf6776f220951917dfa14/common/src/ledger.rs#L78) in `oxidecomputer/omicron` as a test, this PR halved my completion times—it's now 454ms cold and 281ms warm. Before this PR, it was 808ms cold and 579ms warm.
- Inline some of the method candidate checks into `is_valid_method_candidate` and remove some unnecessary visibility checks. This was suggested by @Veykril in [this comment](https://github.com/rust-lang/rust-analyzer/pull/16498#issuecomment-1929864427).

We filter candidate traits by taking advantage of orphan rules. For additional details, I'll rely on @WaffleLapkin's explanation  [from Zulip](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/Trait.20Checking/near/420942417): 

> A type `A` can only implements traits which
> 1. Have a blanket implementation (`impl<T> Trait for T {}`)
> 2. Have implementation for `A` (`impl Trait for A {}`)
> 
> Blanket implementation can only exist in `Trait`'s crate. Implementation for `A` can only exist in `A`'s or `Trait`'s crate.

Big thanks to Waffle for its keen observation!

---

I think some additional improvements are possible:
- `for_trait_and_self_ty` seemingly does not distinguish between `&T`, `&mut T`, or `T`, resulting in seemingly irrelevant traits like `tokio::io::AsyncWrite` being being included for, e.g., `&slog::Logger`. I don't know they're being considered due to the [autoref/autoderef behavior](https://github.com/rust-lang/rust-analyzer/blob/a02a219773629686bd8ff123ca1aa995fa50d976/crates/hir-ty/src/method_resolution.rs#L945-L962), but I wonder if it'd make sense to filter by mutability earlier and not consider trait implementations that require `&mut T` when we only have a `&T`.
- The method completions [spend a _lot_ of time in unification](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/Trait.20Checking/near/421072356), and while there might be low-hanging fruit there, it might make more sense to wait for the new trait solver in `rustc`. I dunno.

[^1]: The filtering occurs outside of typechecking, after all.

WaffleLapkin commented on crates/ide-db/src/imports/import_assets.rs:
```diff
@@ -536,6 +537,29 @@ fn trait_applicable_items(
         Some(assoc_item_trait.into())
     })
+    .filter(|candidate_trait_id| {
+        // blanket `Trait` implementations for type `A` can only exist in the crate defining
+        // the trait. However, an implementation for `A`` can only exist in `A`'s or `Trait`'s crate,
+        // which allows us to reduce the search space substantially.
```
As I later pointer out on zulip, this is not correct for fundamental types. Example (please add something similar as a test):

```rust
struct A;

impl fmt::Display for Box<A> { ... }
impl fmt::Display for &A { ... }
```
Both of those impls [compile](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4e9622b393ae5e8aabbb93133e169703) even though the impls are not in `Display` or `Box`/`&`'s definition crate.

Similarly: 
```rust
struct A;

impl Into<A> for () {
    fn into(self) -> A { A }
}
```
Is also valid, even though both `Into` and unit are foreign.

You need to go through all generics.

WaffleLapkin replied:
Note though that `impl fmt::Display for Vec<A> {  }` does **not** compile. So the crates to search for impls are
1. Trait's definition crate
2. Definition crates for all trait's generic arguments
    - This is recursive for _fundamental types_ — `Into<Box<A>> for ()` is OK, but `Into<Vec<A>> for ()` is NOT
3. Self type definition crate
     - This is recursive for _fundamental types_

davidbarsky replied:
i'l add add a test!

davidbarsky replied:
Wrote my own trait (`pub trait TestInto<T> { fn into(self) -> T; }`, following your suggestion on Zulip) and completions on `()` showed up.

Veykril commented on crates/hir-ty/src/method_resolution.rs:
```diff
@@ -1167,10 +1167,13 @@ fn iterate_trait_method_candidates(
         let mut known_implemented = false;
-        for &(_, item) in data.items.iter() {
+        'items: for &(_, item) in data.items.iter() {
+            if let AssocItemId::TypeAliasId(_) = item {
+                continue 'items;
+            }
```
This is unnecessary, `is_valid_method_candidate` will immediately bail out already

davidbarsky replied:
done!

Veykril commented on crates/hir-ty/src/method_resolution.rs:
```diff
@@ -1459,6 +1533,7 @@ fn is_valid_fn_candidate(
 
         check_that!(table.unify(&expect_self_ty, self_ty));
 
+        let _p = tracing::span!(tracing::Level::INFO, "check_receiver_ty").entered();
```
Wouldn't it make more sense to have these two spans inside the respective following if expressions instead of in front?

davidbarsky replied:
yup, it would.

Veykril commented on crates/hir/src/lib.rs:
```diff
@@ -266,6 +266,10 @@ impl Crate {
         data.potential_cfg_options.clone().unwrap_or_else(|| data.cfg_options.clone())
     }
+
+    pub fn id(&self) -> CrateId {
+        self.id
+    }
```
There are `From` impls to go from hir type to id, no need for this

davidbarsky replied:
removed!

Veykril commented on crates/hir/src/lib.rs:
```diff
@@ -3691,7 +3695,7 @@ pub enum CaptureKind {
 pub struct Type {
     env: Arc<TraitEnvironment>,
-    ty: Ty,
+    pub ty: Ty,
```
This should stay private, we don't want to leak hir-ty `Ty`pes into the ide layer

davidbarsky replied:
removed!

Veykril commented on crates/ide-db/src/imports/import_assets.rs:
```diff
@@ -536,6 +537,32 @@ fn trait_applicable_items(
+        // the trait. However, an implementation for `A`` can only exist in `A`'s or `Trait`'s crate,
+        // which allows us to reduce the search space substantially.
+        let candidate_trait_id: TraitId = *candidate_trait_id;
+        let definining_crate_for_trait = db.trait_environment(candidate_trait_id.into()).krate;
```
```suggestion
        let definining_crate_for_trait = Trait::from(candidate_trait_id).krate();
```

Veykril commented on crates/ide-db/src/imports/import_assets.rs:
```diff
@@ -536,6 +537,32 @@ fn trait_applicable_items(
+        };
+        let definitions_exist_in_trait_crate = db
+            .trait_impls_in_crate(definining_crate_for_trait)
+            .for_trait_and_self_ty(candidate_trait_id, receiver_fingerprint)
```
Might be nice to add a method `has_impls_for_trait_and_self_ty(..) -> bool`

davidbarsky replied:
done!

lnicola commented on crates/hir-ty/src/method_resolution.rs:
```diff
@@ -1414,6 +1417,75 @@ fn is_valid_candidate(
+    let db = table.db;
+    match item {
+        AssocItemId::FunctionId(fn_id) => {
+            let db = table.db;
```
Nit: this is unused (there's another one above).

davidbarsky replied:
whoops, you're right. removed.

davidbarsky commented:
Right, okay: I addressed the review feedback, except the big one from Waffle. I'm pretty sure I've got a handle on checking fundamental types (I'll pilfer from it from the orphan implementation check), but I'm struggling to get a test with fundamental types working (see the changes in `crates/ide-completion/src/tests/flyimport.rs`). The expect side of things isn't the correct correct, but I can't seem to get the fixtures to be aware of a `#[fundamental]` `Box`—any tips?

Veykril commented:
Aren't you missing the fundamental check?

davidbarsky commented:
I am, yeah. I just want to write some tests that invoke fundamental trait impls and then ensure the impl works afterwards (in this PR, to be clear).

Veykril commented:
Oh wait, your test fixture is not using the fundamental type in any impl, so how should it show anything?

davidbarsky commented:
Maybe a better of phrasing my question is "do I need to do anything special to introduce a new fundamental type in a fixture, or is the `#[fundamental]` annotation sufficient". Looking at some of the other tests, I'm inclined to say no.

Veykril commented:
Just using the attribute is enough yes

davidbarsky commented:
Alright, I updated the PR to include tests for `impl ForeignTrait for Box<LocalStruct>`, `impl ForeignTrait for &LocalStruct`, and `impl Into<A> for ()`. The first two tests pass without issue (in that the filtering already considers fundamental types), but the third test seemingly _does not_ work, despite `.into()` showing up in completions when I build a rust-analyzer binary with these exact changes. 

Notably, I _didn't_ need to make additional changes when filtering to consider fundamental types.

davidbarsky commented on crates/ide-completion/src/tests/flyimport.rs:
```diff
@@ -374,6 +374,129 @@ fn main() {
+}
+
+fn main() {
+    let a = A;
```
I don't believe it is, based on Waffle's comment: https://github.com/rust-lang/rust-analyzer/pull/16555#discussion_r1488545233. 

Note that this test also passes _without_ my filtering changes in `crates/ide-db/src/imports/import_assets.rs`, so I'm inclined to investigate this detail later.

Veykril replied:
Well the impl is valid sure, but the completion can't possibly trigger since in your snippet `Into` is implemented for `()`, yet you try to call it on `A` which does not implement `Into`

Veykril replied:
That is `A.into()` doesn't resolve due to the lack of an impl, `<()>.into()` should resolve though as that impl exists

Veykril replied:
Oh, now I see what you mean, right `into` will appears appear given the blanket impls. So any sized type always has at least one applicable `Into` impl

Veykril commented on crates/ide-db/src/imports/import_assets.rs:
```diff
@@ -536,6 +537,29 @@ fn trait_applicable_items(
+        //     a. This is recursive for fundamental types: `Into<Box<A>> for ()`` is OK, but `Into<Vec<A>> for ()`` is *not*.
+        // 3. Receiver type definition crate
+        //    a. This is recursive for fundamental types
+        let candidate_trait_id: TraitId = *candidate_trait_id;
```
(style nit) Could be replaced by changing the closure param to `|&candidate_trait_id|`

davidbarsky replied:
done!

Veykril commented on crates/ide-db/src/imports/import_assets.rs:
```diff
@@ -536,6 +537,29 @@ fn trait_applicable_items(
+        else {
+            return false;
+        };
+        let definitions_exist_in_trait_crate = db
+            .trait_impls_in_crate(defining_crate_for_trait.into())
+            .has_impls_for_trait_and_self_ty(candidate_trait_id, receiver_fingerprint);
+
+        let definitions_exist_in_receiver_crate = db
+            .trait_impls_in_crate(trait_candidate.receiver_ty.krate(db).into())
+            .has_impls_for_trait_and_self_ty(candidate_trait_id, receiver_fingerprint);
+
+        definitions_exist_in_trait_crate || definitions_exist_in_receiver_crate
```
I doubt the compiler will optimize this to be lazy due to db accesses, so making the latter a closure to be called in the or expression would a bit better perf wise

davidbarsky replied:
oh, good point. will fix.

davidbarsky replied:
done!

Veykril commented on crates/hir/src/lib.rs:
```diff
@@ -4239,6 +4239,20 @@ impl Type {
         }
     }
 
+    pub fn is_fundamental(&self, db: &dyn HirDatabase) -> bool {
+        match self.ty.kind(Interner) {
+            TyKind::Ref(_, _, _) => true,
+            &TyKind::Adt(hir_ty::AdtId(AdtId::StructId(s)), ref _subs) => {
+                db.struct_data(s).flags.contains(StructFlags::IS_FUNDAMENTAL)
+            }
+            _ => false,
+        }
+    }
+
+    pub fn fingerprint_for_trait_impl(&self) -> Option<TyFingerprint> {
+        TyFingerprint::for_trait_impl(&self.ty)
+    }
+
```
These aren't needed anymore I think?

davidbarsky replied:
`is_fundamental` is not needed, but `fingerprint_for_trait_impl` is. `.has_impls_for_trait_and_self_ty()` is defined in `hir-ty`, but `self` (`hir::Type`) is defined in `hir`, and I think you can imagine why I'm not keen on figuring out how to break that circular dependency 😅

davidbarsky replied:
cool! just removed `is_fundamental` and pushed.

Veykril commented on crates/hir/src/lib.rs:
```diff
@@ -40,7 +40,7 @@ use std::{iter, mem::discriminant, ops::ControlFlow};
 use base_db::{CrateDisplayName, CrateId, CrateOrigin, Edition, FileId};
 use either::Either;
-use hir_def::{
+pub use hir_def::{
```
```suggestion
use hir_def::{
```

Veykril commented on crates/hir/src/lib.rs:
```diff
@@ -60,7 +60,7 @@ use hir_def::{
 };
 use hir_expand::{attrs::collect_attrs, name::name, proc_macro::ProcMacroKind, MacroCallKind};
-use hir_ty::{
+pub use hir_ty::{
```
```suggestion
use hir_ty::{
```

Veykril commented on crates/hir/src/lib.rs:
```diff
@@ -110,7 +110,7 @@ pub use {
     hir_def::{
         attr::{builtin::AttributeTemplate, AttrSourceMap, Attrs, AttrsWithOwner},
-        data::adt::StructKind,
+        data::adt::{StructFlags, StructKind},
```
```suggestion
        data::adt::StructKind,
```

Veykril commented on crates/hir/src/lib.rs:
```diff
@@ -4239,6 +4239,10 @@ impl Type {
         }
     }
 
+    pub fn fingerprint_for_trait_impl(&self) -> Option<TyFingerprint> {
```
I guess for the changes we need to re-export `TyFingerprint` still? If we do that should go in the big re-export use statement

davidbarsky replied:
weirdly enough, we don't need to reexport `TyFingerprint` for this method. not sure why!

Veykril commented:
Lgtm with the imports tidied up
@bors delegate+

bors commented:
:v: @davidbarsky, you can now approve this pull request!

If @Veykril told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@Veykril`
<!-- homu: {"type":"Delegated","delegator":"Veykril","delegate":"davidbarsky"} -->

Veykril commented on crates/ide-db/src/imports/import_assets.rs:
```diff
@@ -536,6 +537,31 @@ fn trait_applicable_items(
         required_assoc_items.insert(assoc);
         Some(assoc_item_trait.into())
     })
+    .filter(|&candidate_trait_id| {
```
Oh actually, one more thing, I think we can get the same traits multiple times as this point? Especially for an empty fuzzy string search we'll end up with a trait N times (N being its number of assoc items). So maybe we should dedup / collect into a set once first, then filter out by this?

davidbarsky replied:
done!

WaffleLapkin commented on crates/ide-db/src/imports/import_assets.rs:
```diff
@@ -536,6 +537,31 @@ fn trait_applicable_items(
+    .filter(|&candidate_trait_id| {
+        // we care about the following cases:
+        // 1. Trait's definition crate
+        // 2. Definition crates for all trait's generic arguments
+        //     a. This is recursive for fundamental types: `Into<Box<A>> for ()`` is OK, but
+        //        `Into<Vec<A>> for ()`` is *not*.
+        // 3. Receiver type definition crate
+        //    a. This is recursive for fundamental types
```
How does this implement the recursive parts? It seems from the tests that it works, yet I'm struggling to see where you check the crates. I only see checks for two crates here, not the recursive cases...

davidbarsky replied:
> How does this implement the recursive parts?

I'm not entirely sure! I found that the tests just worked with the expected behavior? I assume this occurs due to the implementation of `TyFingerprint` inside of `hir_ty`, but to be entirely honest, I don't fully understand that code yet.

Just to confirm, when you say recursive, are you thinking of a case like:

```rust
impl foreign::TestTrait for fundamental::Box<fundamental::Box<local::TestStruct>> {
    fn some_method(&self) {}
}
```

...where the method receiver is a double-Boxed `TestStruct`?

WaffleLapkin replied:
Yes, that's what I mean. Although you don't even need double boxes, just a single `Box<TestStruct>` will make you check 3 crates (foreign, fundamental, local)...

davidbarsky replied:
I think that's been implemented in the test named `trait_method_fuzzy_completion_aware_of_fundamental_boxes` :)

WaffleLapkin replied:
I know, thus 
> It seems from the tests that it works

but I can't see why

davidbarsky replied:
ah, i misunderstood, sorry! I interpreted " I only see checks for two crates here, not the recursive cases..." as "I don't see tests for the recursive case", not "it's not obvious to me as to why this works".

davidbarsky commented on crates/ide-completion/src/tests/flyimport.rs:
```diff
@@ -374,6 +374,135 @@ fn main() {
+}
+
+fn main() {
+    let t = &TestStruct;
```
agreed! weirdly enough, that test fails, despite a rust-analyzer binary working in this case. not sure why, i'm digging, but i don't get an answer shortly, i'm inclined to figure that out later.

WaffleLapkin replied:
Given that the completion currently works, I don't think we should break it in the name of perf.

Veykril replied:
Might be because you need `//- minicore: deref`, not sure but could be that autoderef exits early without the lang item present

davidbarsky replied:
`//- minicore: deref` didn't work, unfortunately :/. I'm inclined to treat this variant as *not* a blocker and investigate later. Here's my summary:
- `impl ForeignTrait for LocalStruct` and `impl ForeignTrait for &LocalStruct` provide completions on `let t = &TestStruct` in _both_ tests and a running binary.
- `impl ForeignTrait for &LocalStruct` only provides completions for provide completions on `let t = TestStruct` in a running binary. 
    - This makes me think that there's a bug in `minicore`—or that I might not be understanding some important details—because this behavioral split reproduces with and without my filtering changes.

davidbarsky replied:
Made an issue: https://github.com/rust-lang/rust-analyzer/issues/16685

davidbarsky commented:
@bors r=@Veykril

bors commented:
:pushpin: Commit c246a930465b32176dcd2f01ece81996606544eb has been approved by `Veykril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust-analyzer) for this repository.

<!-- @bors r=Veykril c246a930465b32176dcd2f01ece81996606544eb -->
<!-- homu: {"type":"Approved","sha":"c246a930465b32176dcd2f01ece81996606544eb","approver":"Veykril","queue":"https://bors.rust-lang.org/homu/queue/rust-analyzer"} -->

bors commented:
:hourglass: Testing commit c246a930465b32176dcd2f01ece81996606544eb with merge 5fead606bc88d9ed3cb1109bea9c1cb1c42b4ac0...
<!-- homu: {"type":"BuildStarted","head_sha":"c246a930465b32176dcd2f01ece81996606544eb","merge_sha":"5fead606bc88d9ed3cb1109bea9c1cb1c42b4ac0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust-analyzer/actions/runs/8053537238/job/21996579326)
Approved by: Veykril
Pushing 5fead606bc88d9ed3cb1109bea9c1cb1c42b4ac0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Veykril","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust-analyzer/actions/runs/8053537238/job/21996579326"},"merge_sha":"5fead606bc88d9ed3cb1109bea9c1cb1c42b4ac0"} -->
